src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,328 @@
+
+
+
+    private void upgradeFiles(ShardId shard, ShardPath targetPath, final Path targetDir, String folderName, Path[] paths) throws IOException {
+        List<Path> movedFiles = new ArrayList<>();
+        for (Path path : paths) {
+            if (path.equals(targetPath.getDataPath()) == false) {
+                final Path sourceDir = path.resolve(folderName);
+                if (Files.exists(sourceDir)) {
+                    logger.info("{} upgrading {} path {} ", shard, folderName, sourceDir);
+                    try (DirectoryStream<Path> stream = Files.newDirectoryStream(sourceDir)) {
+                        Files.createDirectories(targetDir);
+                        for (Path file : stream) {
+                            if (IndexWriter.WRITE_LOCK_NAME.equals(file.getFileName().toString())) {
+                                continue; // skip write.lock
+                            }
+                            logger.info("{} move file [{}] size: [{}]", shard, file.getFileName(), Files.size(file));
+                            final Path targetFile = targetDir.resolve(file.getFileName());
+                            Files.move(file, targetFile);
+                            movedFiles.add(targetFile);
+                        }
+                    }
+                }
+            }
+        }
+        if (movedFiles.isEmpty() == false) {
+            // fsync later it might be on disk already
+            logger.info("{} fsync files", shard);
+            for (Path moved : movedFiles) {
+                logger.info("{} syncing {}", shard, moved.getFileName());
+                IOUtils.fsync(moved, false);
+            }
+            logger.info("{} syncing directory {}", shard, targetDir);
+            IOUtils.fsync(targetDir, true);
+        }
+    }
+
+
+    /**
+     * Returns <code>true</code> iff the target path is one of the given paths.
+     */
+    private boolean isTargetPathConfigured(final Path[] paths, ShardPath targetPath) {
+        for (Path path : paths) {
+            if (path.equals(targetPath.getDataPath())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Runs an upgrade on all shards located under the given node environment
+     */
+    public static void upgradeMultiDataPath(NodeEnvironment nodeEnv) throws IOException {
+        MultiDataPathUpgrader upgrader = new MultiDataPathUpgrader(nodeEnv);
+        final Set<ShardId> allShardIds = nodeEnv.findAllShardIds();
+        for (ShardId shardId : allShardIds) {
+            try (ShardLock lock = nodeEnv.shardLock(shardId, 0)) {
+                if (upgrader.needsUpgrading(shardId, ImmutableSettings.EMPTY)) { // nocommit - where are we getting the custom paths from?
+                    final ShardPath shardPath = upgrader.pickShardPath(shardId, ImmutableSettings.EMPTY);
+                    upgrader.upgrade(shardId, shardPath);
+                    if (Files.exists(shardPath.resolveIndex())) {
```

Comment:
Shouldn't we throw an exc here if it doesn't exist?  We just upgraded it ... it must exist right?